javalib/src/main/scala/java/util/LinkedHashMap.scala

```diff
@@ -12,72 +12,160 @@
 
 package java.util
 
-import scala.collection.mutable
+import java.{util => ju}
 
-class LinkedHashMap[K, V] private (inner: mutable.LinkedHashMap[Box[K], V],
-    accessOrder: Boolean) extends HashMap[K, V](inner) {
+class LinkedHashMap[K, V](initialCapacity: Int, loadFactor: Float,
+    accessOrder: Boolean)
+    extends HashMap[K, V](initialCapacity, loadFactor) {
   self =>
 
-  def this() =
-    this(mutable.LinkedHashMap.empty[Box[K], V], false)
-
-  def this(initialCapacity: Int, loadFactor: Float, accessOrder: Boolean) = {
-    this(mutable.LinkedHashMap.empty[Box[K], V], accessOrder)
-    if (initialCapacity < 0)
-      throw new IllegalArgumentException("initialCapacity < 0")
-    else if (loadFactor < 0.0)
-      throw new IllegalArgumentException("loadFactor <= 0.0")
-  }
+  import LinkedHashMap._
+
+  /** Node that was least recently created (or accessed under access-order). */
+  private var eldest: Node[K, V] = _
+
+  /** Node that was most recently created (or accessed under access-order). */
+  private var youngest: Node[K, V] = _
 
   def this(initialCapacity: Int, loadFactor: Float) =
     this(initialCapacity, loadFactor, false)
 
   def this(initialCapacity: Int) =
-    this(initialCapacity, LinkedHashMap.DEFAULT_LOAD_FACTOR)
+    this(initialCapacity, HashMap.DEFAULT_LOAD_FACTOR)
+
+  def this() =
+    this(HashMap.DEFAULT_INITIAL_CAPACITY)
 
   def this(m: Map[_ <: K, _ <: V]) = {
-    this()
+    this(m.size())
     putAll(m)
   }
 
-  override def get(key: scala.Any): V = {
-    val value = super.get(key)
-    if (accessOrder) {
-      val boxedKey = Box(key.asInstanceOf[K])
-      if (value != null || containsKey(boxedKey)) {
-        inner.remove(boxedKey)
-        inner(boxedKey) = value
+  private def asMyNode(node: HashMap.Node[K, V]): Node[K, V] =
+    node.asInstanceOf[Node[K, V]]
+
+  private[util] override def newNode(key: K, hash: Int, value: V,
+      previous: HashMap.Node[K, V],
+      next: HashMap.Node[K, V]): HashMap.Node[K, V] = {
+    new Node(key, hash, value, previous, next, null, null)
+  }
+
+  // accepts `node eq null`
+  private[util] override def touchNode(node: HashMap.Node[K, V]): Unit = {
+    if (accessOrder && (node ne null)) {
+      val myNode = asMyNode(node)
+      val younger = myNode.younger
+      if (younger ne null) {
+        // Remove myNode from its previous spot
+        val older = myNode.older
+        younger.older = older
+        if (older ne null)
+          older.younger = younger
+        else
+          eldest = younger
+
+        // Insert myNode at the end of the list
+        val oldYoungest = youngest
+        oldYoungest.younger = myNode
+        myNode.older = oldYoungest
+        myNode.younger = null
+        youngest = myNode
```

Comment:
Could you create two helpers for `insertYounger` and `removeSomething`? I think it will help readability quite a bit.

Reply:
Helpers that will be used only from `touchNode` (renamed `nodeWasAccessed`)? Would that be more readable than the comments for the two sections? I'm not sure I understand what you mean here, could you elaborate?

Reply:
Ah, sorry. I meant one helper to add a node to the iteration list and one helper to remove it. You can then also use these from `nodeWasAdded`, `nodeWasRemoved`.

Reply:
Ah, OK. Done.